<a href="https://colab.research.google.com/github/sebekpro/python-krakowiak/blob/main/250-data-science-bootcamp-w-jezyku-python/211_220_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Science Bootcamp

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 211](#1)
* [Ćwiczenie 212](#2)
* [Ćwiczenie 213](#3)
* [Ćwiczenie 214](#4)
* [Ćwiczenie 215](#5)
* [Ćwiczenie 216](#6)
* [Ćwiczenie 217](#7)
* [Ćwiczenie 218](#8)
* [Ćwiczenie 219](#9)
* [Ćwiczenie 220](#10)

### <a name='0'></a>Import biblioteki

In [ ]:
import numpy as np
import pandas as pd

np.__version__

'2.0.2'

### <a name='1'></a> Ćwiczenie 211
Podany jest poniższy obiekt _DataFrame_:

```
df = pd.DataFrame(data={'height': [175., 178.5, 185., 191., 184.5, 183., 168.]})
```
Dokonaj dyskretyzacji zmiennej _height_ na 3 przedziały o równej szerokości. Wynik przypisz do nowej kolumny _height_cut_.


In [ ]:
# tutaj wpisz rozwiązanie
df = pd.DataFrame(data={'height': [175., 178.5, 185., 191., 184.5, 183., 168.]})

# Dyskretyzacja zmiennej 'height' na 3 przedziały o równej szerokości
df['height_cut'] = pd.cut(df['height'], bins = 3)

df

height          height_cut
0   175.0  (167.977, 175.667]
1   178.5  (175.667, 183.333]
2   185.0    (183.333, 191.0]
3   191.0    (183.333, 191.0]
4   184.5    (183.333, 191.0]
5   183.0  (175.667, 183.333]
6   168.0  (167.977, 175.667]

Funkcja pd.cut dzieli ciągły zakres wartości zmiennej na określoną liczbę przedziałów o równej szerokości. W przypadku zmiennej height oraz argumentu bins=3:

1. Obliczanie zakresu wartości:
Funkcja wyznacza minimalną i maksymalną wartość w kolumnie height. Dla przykładu, jeśli mamy wartości od 168.0 do 191.0, to zakres wynosi 23.0.

2. Wyznaczenie szerokości przedziału:
Całkowity zakres jest dzielony przez liczbę przedziałów. W naszym przykładzie szerokość jednego przedziału wynosi 23.0/0 ≈ 7.67

3. Ustalenie granic przedziałów:
Na podstawie minimalnej wartości oraz szerokości przedziału funkcja automatycznie określa granice:
* Pierwszy przedział: od 168.0 do 168.0 + 7.67 ≈ 175.67,
* Drugi przedział: od 175.67 do 175.67 + 7.67 ≈ 183.33,
* Trzeci przedział: od 183.33 do 191.0.

Domyślnie przedziały są tworzone jako otwarte z lewej i domknięte z prawej stroną, czyli np. (168.0, 175.67].

4. Przypisanie wartości do przedziałów:
Każda wartość w kolumnie height jest przyporządkowywana do odpowiedniego przedziału, w którym się znajduje. W wyniku tej operacji otrzymujemy zmienną kategoryczną, gdzie każda obserwacja ma etykietę przedziału (np. (175.67, 183.33]).

Podsumowując, dyskretyzacja zmiennej height przy użyciu pd.cut(height, bins=3) polega na automatycznym podziale zakresu wartości na 3 równe segmenty, a następnie przypisaniu każdej wartości do odpowiadającego jej segmentu. Wynik jest nową kolumną, która kategoryzuje wartości zmiennej height według ustalonych przedziałów.

### <a name='2'></a> Ćwiczenie 212
Dokonaj dyskretyzacji zmiennej _height_ na 3 przedziały o zadanej postaci:
* (160, 175]
* (175, 180]
* (180, 195]

Wynik przypisz do nowej kolumny _height_cut_2_.

In [ ]:
# tutaj wpisz rozwiązanie

# Definicja przedziałów
bins = [160, 175, 180, 195]

# Dyskretyzacja zmiennej 'height' przy użyciu zdefiniowanych przedziałów
df['height_cut_2'] = pd.cut(df['height'], bins = bins)

df

height          height_cut height_cut_2
0   175.0  (167.977, 175.667]   (160, 175]
1   178.5  (175.667, 183.333]   (175, 180]
2   185.0    (183.333, 191.0]   (180, 195]
3   191.0    (183.333, 191.0]   (180, 195]
4   184.5    (183.333, 191.0]   (180, 195]
5   183.0  (175.667, 183.333]   (180, 195]
6   168.0  (167.977, 175.667]   (160, 175]

1. Definicja przedziałów:
Lista bins = [160, 175, 180, 195] określa granice przedziałów. Funkcja pd.cut() domyślnie tworzy przedziały w postaci (a, b] – czyli lewy brzeg jest otwarty, a prawy domknięty. W efekcie otrzymamy przedziały:
* (160, 175]
* (175, 180]
* (180, 195]
2. Dyskretyzacja:
Funkcja pd.cut(df['height'], bins=bins) przypisuje każdą wartość zmiennej height do odpowiedniego przedziału.

### <a name='3'></a> Ćwiczenie 213
Dokonaj dyskretyzacji zmiennej _height_ na 3 przedziały o zadanej postaci:
* (160, 175]
* (175, 180]
* (180, 195]

oraz przypisz im odpowiednio etykiety:

* small
* medium
* high

Wynik przypisz do nowej kolumny _height_cut_3_.

In [ ]:
# tutaj wpisz rozwiązanie

# Definicja przedziałów oraz etykiet
bins = [160, 175, 180, 195]
labels = ['small', 'medium', 'height']

# Dyskretyzacja zmiennej 'height' z przypisaniem etykiet
df['height_cut_3'] = pd.cut(df['height'], bins = bins, labels = labels)

df

height          height_cut height_cut_2 height_cut_3
0   175.0  (167.977, 175.667]   (160, 175]        small
1   178.5  (175.667, 183.333]   (175, 180]       medium
2   185.0    (183.333, 191.0]   (180, 195]       height
3   191.0    (183.333, 191.0]   (180, 195]       height
4   184.5    (183.333, 191.0]   (180, 195]       height
5   183.0  (175.667, 183.333]   (180, 195]       height
6   168.0  (167.977, 175.667]   (160, 175]        small

1. **Definicja przedziałów i etykiet:**
- **Lista bins** = `[160, 175, 180, 195]` określa granice przedziałów, które domyślnie tworzą przedziały w postaci (a, b].
- *Lista labels* = `['small', 'medium', 'high']` zawiera etykiety przypisane kolejno do przedziałów:
  - (160, 175] – small
  - (175, 180] – medium
  - (180, 195] – high
2. Dyskretyzacja z pd.cut():
Funkcja pd.cut() dzieli zmienną height według zdefiniowanych granic, a parametr labels=labels powoduje, że każdemu przedziałowi przypisywana jest odpowiednia etykieta.

### <a name='4'></a> Ćwiczenie 214
Usuń kolumny _height_cut_ oraz _height_cut_2_ z obiektu _df_.

In [ ]:
# tutaj wpisz rozwiązanie
df = df.drop(columns = ['height_cut', 'height_cut_2'])
df

height height_cut_3
0   175.0        small
1   178.5       medium
2   185.0       height
3   191.0       height
4   184.5       height
5   183.0       height
6   168.0        small

Zmień nazwę kolumny _height_cut_3_ na _height_level_.

In [ ]:
# tutaj wpisz rozwiązanie
df = df.rename( columns = {'height_cut_3': 'height_level'})
df

height height_level
0   175.0        small
1   178.5       medium
2   185.0       height
3   191.0       height
4   184.5       height
5   183.0       height
6   168.0        small

### <a name='5'></a> Ćwiczenie 215
Uzyskaj kodowanie 0-1 obiektu _df_ dzięki funkcji _pd.get_dummies()_.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='6'></a> Ćwiczenie 216
Podany jest poniższy obiekt _df_:


```
df = pd.DataFrame(data={'lang': [['PL', 'ENG'], ['GER', 'ENG', 'PL', 'FRA'], ['RUS']]})
```
Zbuduj nową kolumnę o nazwie _length_, która przyjmie długość listy kolumny _lang_.



In [ ]:
# tutaj wpisz rozwiązanie

df_lang = pd.DataFrame(data={'lang': [['PL', 'ENG'], ['GER', 'ENG', 'PL', 'FRA'], ['RUS']]})
df_lang['length'] = df_lang['lang'].apply(len)
df_lang

lang  length
0            [PL, ENG]       2
1  [GER, ENG, PL, FRA]       4
2                [RUS]       1

- df['lang'].apply(len): Dla każdej wartości w kolumnie lang (która jest listą) funkcja len zwraca liczbę elementów w liście.

### <a name='7'></a> Ćwiczenie 217
Zbuduj nową kolumnę o nazwie _PL_flag_, która przyjmie wartość 1, gdy język 'PL' będzie w liście w kolumnie _lang_, przeciwnie 0.

In [ ]:
# tutaj wpisz rozwiązanie

df_lang['PL_flag'] = df_lang['lang'].apply(lambda langs: 1 if 'PL' in langs else 0)

df_lang

lang  length  PL_flag
0            [PL, ENG]       2        1
1  [GER, ENG, PL, FRA]       4        1
2                [RUS]       1        0

- df['lang'].apply(lambda langs: 1 if 'PL' in langs else 0):
Dla każdej listy języków w kolumnie lang funkcja lambda sprawdza, czy element 'PL' znajduje się w tej liście. Jeśli tak – zwraca 1, w przeciwnym razie 0.

### <a name='8'></a> Ćwiczenie 218
Podany jest poniższy obiekt _df_:


```
df = pd.DataFrame(data={'website': ['wp.pl', 'onet.pl', 'google.com']})
```
Podziel kolumnę _website_ względem znaku kropki (.) na dwie kolumny o nazwach _portal_ oraz _ending_.


In [ ]:
# tutaj wpisz rozwiązanie
df_web = pd.DataFrame(data={'website': ['wp.pl', 'onet.pl', 'google.com']})

df_web[['portal', 'domain']] = df_web['website'].str.split(r'\.', expand = True)

df_web

website  portal domain
0       wp.pl      wp     pl
1     onet.pl    onet     pl
2  google.com  google    com

- **df['website'].str.split(r'.', expand=True):**
Funkcja str.split() dzieli wartości w kolumnie website względem znaku kropki. Używamy surowego ciągu znaków (r'\.'), aby kropka była traktowana dosłownie (względem wyrażeń regularnych).
- **expand=True:**
Dzięki temu wynik dzielenia jest rozwijany do DataFrame, w którym każda kolumna odpowiada jednemu przedziałowi wynikowemu.

### <a name='9'></a> Ćwiczenie 219
Wytnij wszystkie wiersze obiektu _df_ dla którego kolumna _portal_ zawiera literę 'o'.

In [ ]:
# tutaj wpisz rozwiązanie
df_with_o = df_web[df_web['portal'].str.contains('o')]
df_with_o

website  portal domain
1     onet.pl    onet     pl
2  google.com  google    com

- df['portal'].str.contains('o'): Tworzy maskę logiczną, która zwraca True dla tych wierszy, gdzie wartość w kolumnie portal zawiera literę 'o'.
- df[...]: Używając tej maski, wybieramy tylko te wiersze, które spełniają warunek.

### <a name='10'></a> Ćwiczenie 210
Zamień wszystkie rozszerzenia _com_ z kolumny _ending_ na _org_.

In [ ]:
# tutaj wpisz rozwiązanie
df_web['domain'] = df_web['domain'].replace('com', 'org')

df_web

website  portal domain
0       wp.pl      wp     pl
1     onet.pl    onet     pl
2  google.com  google    org

- df['ending'].replace('com', 'org'):
Metoda replace() przeszukuje kolumnę ending i zamienia wszystkie wystąpienia wartości `'com'` na `'org'`.